lxd/device/pci.go

```diff
@@ -0,0 +1,122 @@
+package device
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	pcidev "github.com/lxc/lxd/lxd/device/pci"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/shared/validate"
+)
+
+type pci struct {
+	deviceCommon
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *pci) validateConfig(instConf instance.ConfigReader) error {
+	if !instanceSupported(instConf.Type(), instancetype.VM) {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(string) error{
+		"address": validate.IsPCIAddress,
+	}
+
+	err := d.config.Validate(rules)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to validate config")
+	}
+
+	// PCI devices can be specified as "0000:XX:XX.X" or "XX:XX.X".
+	// However, the devices in /sys/bus/pci/devices use the long format which
+	// is why we need to make sure the prefix is present.
+	if len(d.config["address"]) == 7 {
+		d.config["address"] = fmt.Sprintf("0000:%s", d.config["address"])
+	}
+
+	return nil
+}
+
+// validateEnvironment checks if the PCI device is available.
+func (d *pci) validateEnvironment() error {
+	return validatePCIDevice(d.config)
+}
+
+// Start is run when the device is added to the instance.
+func (d *pci) Start() (*deviceConfig.RunConfig, error) {
+	err := d.validateEnvironment()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to validate environment")
+	}
+
+	runConf := deviceConfig.RunConfig{}
+	saveData := make(map[string]string)
+
+	// Get PCI information about the device.
+	pciAddress := d.config["address"]
+	devicePath := filepath.Join("/sys/bus/pci/devices", pciAddress)
+	pciDev, err := pcidev.ParseUeventFile(filepath.Join(devicePath, "uevent"))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get PCI device info for %q", pciAddress)
+	}
+
+	saveData["last_state.pci.slot.name"] = pciDev.SlotName
```

Comment:
@stgraber do we need to store slot name in volatile for this device (isn't it always the same as the d.config["address"] value?).

Reply:
We probably don't need to but given we do it everywhere else we deal with PCI devices, might as well keep things consistent :)

Reply:
Fair enough :)